### PR TITLE
Rewrite for guild setup system

### LINKED ISSUE #691: Issues with init logic
Currently JDA requires all guilds and all of the associated entities to be loaded before starting to fire events. This has serious effects as entire shards can get stuck on a single guild refusing to connect to the gateway. In addition to that JDA can have serious cache issues during outages that can lead to memory issues due to duplicated cache entries that don't get removed.

- [ ] Guilds should not prevent startup
- [ ] Events should fire when all required entities are loaded
- [ ] Cache should be aware of duplicates
- [ ] Properly handle unavailability

We have ideas of adding a `GuildReadyEvent` in addition to `ReadyEvent` which would be useful to allow events to fire even if not all guilds are ready yet. The `ReadyEvent` would be fired when all guilds have been initialized, however events may already have fired for them.

I will redirect the issue of #625 to this issue because they are closely related to one-another but I will keep it open for now.

As I don't have a lot of free time at the moment, I cannot give any ETA on when I will find time to work on these proposals. This most likely requires a rewrite of JDA internals.

### PR DESCRIPTION
**Pull request by MinnDevelopment**
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

There are several guidelines you should follow in order for your
  Pull Request to be merged.

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

> It is sometimes better to include more changes in a single commit. 
  If you find yourself having an overwhelming amount of commits, you
  can **rebase** your branch.

## Description

Fixes #691, #625, #602, #512 

These changes will make it possible to fire events for guilds without having all guilds loaded. For this I have added `GuildReadyEvent` which will fire when a guild that we received in the `READY` dispatch finished setting up. In most cases this is done as soon as a `GUILD_CREATE` arrives, others will be added for chunking.

These changes are not fully completed at this time, however I want everyone to have a chance to start looking through my changes in order to spot any issues that I might have missed so far.
 
I'm not sure what version I should bump for these changes, in all honesty I'm considering major as it affects every user of the library.

### CODE REVIEW

File: src/main/java/net/dv8tion/jda/core/handle/VoiceServerUpdateHandler.java [hunk 1]
```diff
@@ -36,13 +36,15 @@ public VoiceServerUpdateHandler(JDAImpl api)
...
+        if (api.getGuildSetupController().isLocked(guildId))
+            return guildId;
         Guild guild = api.getGuildMap().get(guildId);
         if (guild == null)
             throw new IllegalArgumentException("Attempted to start audio connection with Guild that doesn't exist! JSON: " + content);
 
         api.getClient().updateAudioConnection(guildId, guild.getSelfMember().getVoiceState().getChannel());
 
-        if (api.getGuildLock().isLocked(guildId))
+        if (api.getGuildSetupController().isLocked(guildId))
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/core/handle/VoiceServerUpdateHandler.java]: this check doesn't make sense

**Comment by MinnDevelopment** [comment]
## Summary of changes

+ Added GuildSetupController/-Node
+ Events will now fire when all entities required for it are ready, not the entire cache
+ Added GuildReadyEvent to see when a guild finishes starting up (only for init, you can still use GuildJoinEvent for real joins)
+ Introduced CacheFlags for more configuration of cache, not all bots need info about user's online status
- ReadyEvent remains unchanged but is no longer the guaranteed first event to fire
- EventCache has been updated and improved, the events are now cached together with meta-information relevant to their behavior and can timeout when in cache for too long

**Comment by MinnDevelopment** [comment]
I have added a status enum to `GuildSetupController` to allow for better debugging. All status changes for nodes will be tracked on `TRACE` logging level for `GuildSetupController`.

`[{GUILD_ID}] Updated status {OLD_STATUS}->{NEW_STATUS}`

Additionally you can now use `GuildSetupController#getSetupNodes(Status)` to see which nodes have which status allowing to do real runtime debugging without having to analyze logs.

**Comment by kantenkugel** [comment]
After looking at the source, I think you have to add some public modifiers to methods in `GuildSetupNode` (and maybe make the class itself public?). Cuz currently they are all default visibility. So `GuildSetupController#getSetupNodes(Status)` is currently not really useful for non-internal code

**Comment by MinnDevelopment** [comment]
The visibility is intentions. `GuildSetupController#getSetupNodes(Status)` is useful to see how many guilds are on that specific status.

**Comment by MinnDevelopment** [comment]
I have now been able to test the edge case of guilds being unavailable during startup and coming back way later, they will properly recover and fire a `GuildReadyEvent` and will not prevent startup to complete for other guilds, the `ReadyEvent` is fired without that specific unavailable guild being ready (as intended).

**Comment by MinnDevelopment** [comment]
After one more week of testing I plan to merge this on the 2nd of August (that's next sunday).
